Making VerifyAllColumn case-insensitive

algirdasN opened the pull request:
### 🤔 What's changed?

Made the comparison case-insensitive in the method that is used to verify that all columns in the table are bound to a member on the object that is being created.

### ⚡️ What's your motivation? 

In order to use VerifyAllColumnsBound option when creating an object instance from a table the field key must match the member's name exactly. This is an inconsistent behavior with instance creation as the members are matched ignoring the case. This makes enabling this option frustrating as one have to go back to feature files and fix all the keys.

### 🏷️ What kind of change is this?

- :zap: New feature (non-breaking change which adds new behaviour)

### 📋 Checklist:

- [x] I've changed the behaviour of the code
  - [x] I have added/updated tests to cover my changes.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.
- [x] Users should know about my change
  - [x] I have added an entry to the "[vNext]" section of the [**CHANGELOG**](../blob/main/CHANGELOG.md), linking to this pull request & included my GitHub handle to the release contributors list.

----

*This text was originally taken from the [template of the Cucumber project](https://github.com/cucumber/.github/blob/main/.github/PULL_REQUEST_TEMPLATE.md), then edited by hand. [You can modify the template here.](https://github.com/reqnroll/.github/edit/main/.github/PULL_REQUEST_TEMPLATE.md)*

304NotModified reviewed (changes requested):
Thanks for the PR. 

But this is a (behavior) breaking change.

I propose to add an option to the options object and add some unit tests. The default behavior should not change.

algirdasN commented:
I was thinking whether this would be breaking and thought that any bug introduced here would already exist since currently Reqnroll will not properly populate instances that have two or more members with matching case-insensitive names.

In any case I've added a property to InstanceCreationOptions and added unit tests to test default and new behavior.

While doing this I've found an interesting behavior of RequireTableToProvideAllConstructorParameters flag. With it unset (default behavior) it will not allow for any valid members that are not used in the constructor (invalid names are ignored and are allowed if not verified). I'm not entirely sure if this is intended because that does not correspond to the name of the flag.

304NotModified commented:
> While doing this I've found an interesting behavior of RequireTableToProvideAllConstructorParameters flag. With it unset (default behavior) it will not allow for any valid members that are not used in the constructor

Could you give an example of this? Thanks!

304NotModified commented on Reqnroll/Assist/InstanceCreationOptions.cs:
```diff
@@ -3,6 +3,7 @@ namespace Reqnroll.Assist
     public class InstanceCreationOptions
     {
         public bool VerifyAllColumnsBound { get; set; }
+        public bool CaseInsensitiveColumnVerify { get; set; }
```
Maybe start with Verify, like the  VerifyAllColumnsBound? (consistency)

e.g.  VerifyCaseInsensitive

304NotModified replied:
All other looks good

304NotModified commented:
Please update also the changelog :)
It's now a feature :)

algirdasN commented on Tests/Reqnroll.RuntimeTests/AssistTests/CreateInstanceHelperMethodTests.cs:
```diff
@@ -179,6 +221,29 @@ public void SetConstructor_unbound_column_throws_ColumnCouldNotBeBoundException_
+        
+        
+        [Fact]
+        public void SetConstructor_column_case_mismatch_throws_ColumnCouldNotBeBoundException_on_verify()
```
> > While doing this I've found an interesting behavior of RequireTableToProvideAllConstructorParameters flag. With it unset (default behavior) it will not allow for any valid members that are not used in the constructor
> 
> Could you give an example of this? Thanks!

This and the following unit tests will fail without RequireTableToProvideAllConstructorParameters flag as the table contains a valid field not in the constructor. If you would look at an old unit test just above this one, it passes with default behavior because the extra field is invalid.

304NotModified reviewed (approved):
Great work!

I also added your username to che changelog

304NotModified commented:
Ow almost forgot,
Could you please update 
https://github.com/reqnroll/Reqnroll/blob/main/docs/automation/datatable-helpers.md?plain=1

See 
https://docs.reqnroll.net/latest/automation/datatable-helpers.html#instancecreationoptions

Thanks a lot!

algirdasN commented:
I've updated the docs. Please let me know if something should be changed. Thank you for guiding me through the process.

gasparnagy commented:
@algirdasN Thank you for the contribution! According to our [guidelines](https://github.com/orgs/reqnroll/discussions/67) I have invited you to the Reqnroll "contributors team". Congrats! 🎉 If you accept it, you will be able to make pull requests easier in the future.

You are also welcome on our discord server: https://go.reqnroll.net/discord-invite
